### PR TITLE
Simplify median filter

### PR DESCRIPTION
**Pull request by tcm0116**
I was having a hard time understanding the implementation for the median filter, so I figured I'd suggest a simpler implementation that is easier to understand.

### CODE REVIEW

**Comment by p3p** [comment]
I agree that the "Ekstrom" Median method is a bit hard to follow, and it could even turn out to be bugged I guess although I've ran a lot of test data through, performance wise it is excellent according to my tests against simpler methods, like std::sort (Quicksort algorithm) and insertion sort even at low values .. do you think I should just use a clearer algorithm?, people seem to still be getting noise through a median filter set to 23 + a 16(ish, doesnt really work like that) sample lowpass? tis boggling my mind

**Comment by tcm0116** [comment]
I was mostly thinking that an easier algorithm might be potentially easier
to maintain and less error prone. However, if you're confident in the
algorithm you've got, then just stick with it.

I think we should stick with 23 values for the filter and move on. If
people are still having issues, then we can re-address.

I did notice I was still getting a bit of oscillation and I saw some on the
screenshots people were providing. I'm wondering of there's something else
going on with the PID beyond the dT issue I found. Do we know for sure that
the temperature ISR is running at 1000Hz? Does it even really need to run
that fast?

**Comment by p3p** [comment]
> Do we know for sure that the temperature ISR is running at 1000Hz?

The ISR runs at 1000Hz but it starts a sample on one pass, then reads it the next, so it gets a sample at 500Hz (so any sort function is running at 500hz), but then it multisamples 16 times, so it gets a full reading at 30Hz. so the filtering actual has a 16x multisample on top of what we are doing. If you want to test your filter to see if you get better results I'l be happy to know, I just didn't want to add to much computation in an ISR, (which was why my original median only had 3 values)

**Comment by tcm0116** [comment]
I'm wondering about `MIN_ADC_ISR_LOOPS` being defined as 10. This makes `PID_dT` be 160 milliseconds. If it's really getting a full reading at 30Hz, then that seems like it should be more like 33 milliseconds.

**Comment by p3p** [comment]
I did a lot of playing with this just to make sure and I'm very confident the filter is performing as expected now, it also profiled at 10x faster than the std::sort method.
(#define ADC_MEDIAN_FILTER_SIZE 31)
Ekstrom + LP: 106075µs
std::sort + LP: 1145515µs
Smoothie: 1216088µs

here are some fun graphs with pretty bad simulated noise on sine waves that are much higher frequency than any temperature change would be:
image1
image2
(#define ADC_MEDIAN_FILTER_SIZE 23)
image3

I'm not sure how I can get any better results.. also my simulated noise may not be the best

I'm not sure about the MIN_ADC_ISR_LOOPS value, you may need to ask someone that understands Marlin better

**Comment by tcm0116** [comment]
Those graphs are really cool.

One thing I've noticed when looking at the OctoPrint temperature graphs (mine included) is that the noise is always downward spikes. Have you tried any simulations with noise like that?

**Comment by p3p** [comment]
Changed the noise to 10% read errors (min/max adc value) 50% chance of random noise at 40% magnitude, the signal always has 5% magnitude noise (normal ADC read deviation)

newimage2
newimage3

The filter is performing better than I would expect tbh, with all the talk of "it works on smoothieware" I thought their filtering would be better, of course it may be PID control on smoothie that is much more resilient to any noise that does slip through.

If we dont care about response time, (and we don't at this frequency) then turning up the Lowpass fiter to 6 has an insane effect for no performance overhead:
newimage4

And even at 70% of 60% magnitude noise:
newimage5

Gotta stop playing with now spent far to much time messing around, although I think I will increase the Lowpass K value to 6 in the PR
 
(one last one for constant signal, temperature holding)
newimage6
